### ISSUE TITLE
fix: Unhandled Exception: 128: Not a supported op.

### ISSUE THREAD

**Issue by nosmirck** [wolfenrain/umbra#37]
**Description**

A simple shader example compiled and generated successfully, however, fails on execution with the following error `Unhandled Exception: 128: Not a supported op.`

**Steps To Reproduce**

Sample shader
```glsl
uniform float time;

const float Pi=3.14159;
const int complexity=40;// More points of color.
const float fluid_speed=3000.;// Drives speed, higher number will make it slower.
const float color_intensity=.225;

vec4 fragment(vec2 uv,vec2 fragCoord){
    vec2 p=(2.*gl_FragCoord.xy-resolution)/max(resolution.x,resolution.y);
    for(int i=1;i<complexity;i++)
    {
        vec2 newp=p+time*.0003;
        newp.x+=.6/float(i)*sin(float(i)*p.y+time/fluid_speed*float(i+2000))+.5;
        newp.y+=.6/float(i)*sin(float(i)*p.x+time/fluid_speed*float(i+1909))-.5;
        p=newp;
    }
    vec3 col=vec3(0.,color_intensity*sin(3.*p.x)+.25,color_intensity*sin(3.*p.x)+color_intensity);
    return vec4(col,1.);
}
```

Flutter code:
```dart
import 'package:flutter/material.dart';
import 'package:shader/shaders/hello_world.dart';

void main() async {
  // Ensure bindings are initialized otherwise we can't user rootBundle.
  WidgetsFlutterBinding.ensureInitialized();

  runApp(const MaterialApp(home: MyApp()));
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  // Compile the binary into a Fragment program.
  var _value = 0.5;
  HelloWorld? program;
  @override
  void initState() {
    super.initState();
    initProgram();
  }

  Future<void> initProgram() async {
    program = await HelloWorld.compile();
    if (mounted) {
      setState(() {});
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            SizedBox(
              width: 300,
              child: Slider(
                value: _value,
                onChanged: (value) {
                  setState(() {
                    _value = value;
                  });
                },
              ),
            ),
            program == null
                ? const CircularProgressIndicator()
                : SizedBox(
                    width: 512,
                    height: 512,
                    child: FlutterShaderTest(
                      shader: program!.shader(
                        resolution: const Size.square(512),
                        time: _value,
                      ),
                    ),
                  ),
          ],
        ),
      ),
    );
  }
}

class FlutterShaderTest extends StatelessWidget {
  const FlutterShaderTest({Key? key, required this.shader}) : super(key: key);

  final Shader shader;

  @override
  Widget build(BuildContext context) {
    return CustomPaint(painter: ShaderPainter(shader));
  }
}

class ShaderPainter extends CustomPainter {
  ShaderPainter(this.shader) : _paint = Paint()..shader = shader;

  final Shader shader;

  final Paint _paint;

  @override
  void paint(Canvas canvas, Size size) {
    canvas.drawRect(const Rect.fromLTWH(0, 0, 512, 512), _paint);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}
```

**Expected Behavior**

Output of the painter should look similar to this example:
https://glslsandbox.com/e#81872.0

**Stacktrace**
```
Launching lib\main.dart on Windows in debug mode...
Connecting to VM Service at ws://127.0.0.1:58165/tK7yafOyS04=/ws
[ERROR:flutter/lib/ui/ui_dart_state.cc(198)] Unhandled Exception: 177: Not a supported op.
#0      _Transpiler.parseInstruction (dart:_spirv/src/transpiler.dart:564:9)
#1      _Transpiler.transpile (dart:_spirv/src/transpiler.dart:169:7)
#2      transpile (dart:_spirv:73:5)
#3      new FragmentProgram._ (dart:ui/painting.dart:3881:44)
#4      FragmentProgram.compile.<anonymous closure> (dart:ui/painting.dart:3872:58)
#5      new Future.<anonymous closure> (dart:async/future.dart:252:37)
#6      _rootRun (dart:async/zone.dart:1418:47)
#7      _CustomZone.run (dart:async/zone.dart:1328:19)
#8      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
#9      _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1276:23)
#10     _rootRun (dart:async/zone.dart:1426:13)
#11     _CustomZone.run (dart:async/zone.dart:1328:19)
#12     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:1260:23)
#13     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
#14     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:398:19)
#15     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:429:5)
#16     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
```

**Additional comments**

I'm not new to shaders, but it's been at least 14 years since the last time I wrote one, back with glsl and GLUT when I was doing my Computer Graphics Specialization. I feel rusty and lost, like it is totally new for me and since Flutter added shader support I've been looking forward to get into this again, at least as a hobby. I found your work (it's amazing!) and I've been following you since that first medium article you wrote.

Today, after a few months of being swamped in work I can finally give it a try and when trying to transcribe one simple example from GLSL Sandbox I get runtime errors that I can't figure out.

I believe that the main issue is that I'm using glsl instructions that don't fully translate to spir-v (for which I have ZERO experience with).

I've also tried to find information about which instructions from glsl are not available or have a different counterpart when translated to spir-v. If you can point me into the right direction, I'd be more than grateful! I don't really know which instruction is a not supported operation, I whish this tool was a bit more explicit on which line of the shader code is having an issue (I know that's almost impossible, but one can dream).

Anyways, thanks again for your amazing work!

**Comment by wolfenrain** [comment]
First of all, thank you for your kind words! And glad you are picking up shader stuff again (and this time in Flutter)! 

To get to your issue, what you are saying is correct. Your code is using the [OpIAdd](https://github.com/KhronosGroup/SPIRV-Headers/blob/b765c355f488837ca4c77980ba69484f3ff277f5/include/spirv/unified1/spirv.lua#L1266) instruction, which is not yet supported by the Flutter SPIRV library.

Sadly there isn't any good documentation on what is implemented and what is not except for this [`constant.dart`](https://github.com/flutter/engine/blob/main/lib/spirv/lib/src/constants.dart#L49) file.

While knowing which exact line is the culprit is indeed a difficult dream to achieve, it might be interesting for Umbra to add a feature where it converts the unknown OP codes to readable string so developers know what "feature" in their GLSL code is at least throwing the OP code exception.

**Comment by nosmirck** [comment]
@wolfenrain that helped a lot!

I found that there were many more things I had in that code that didn't work, for instance, the LessThan `<` is not supported either.

The specific problem with op 128 was the `+=` so, that's an easy fix, but there were more issues, like operating int + float and the for loop (because of the `<`)

Anyways, I managed to get the shader loading, but, it fails with a totally different exception this time (`Unhandled Exception: Invalid SkSL`), it points to the generated glsl file, still, I have no idea where it is.

Here is the updates shader code, as you can see, I removed/changed all the offending ops that are not supported and expanded the loop (just a few times)

```glsl
uniform float time;

const float Pi=3.14159;
//const int complexity=40;// More points of color.
const float fluid_speed=3000.;// Drives speed, higher number will make it slower.
const float color_intensity=.225;

vec2 newP(vec2 p,float time,float i){
    vec2 newp=p+time*.0003;
    newp.x=newp.x+.6/i*sin(i*p.y+time/fluid_speed*i+2000.)+.5;
    newp.y=newp.y+.6/i*sin(i*p.x+time/fluid_speed*i+1909.)-.5;
    return newp;
}

vec4 fragment(vec2 uv,vec2 fragCoord){
    vec2 p=(2.*gl_FragCoord.xy-resolution)/max(resolution.x,resolution.y);
    float i=0.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    i=i+1.;
    p=newP(p,time,i);
    vec3 col=vec3(0.,color_intensity*sin(3.*p.x)+.25,color_intensity*sin(3.*p.x)+color_intensity);
    return vec4(col,1.);
}
```

And this is the StackTrace:

```
Restarted application in 292ms.
[ERROR:flutter/lib/ui/ui_dart_state.cc(198)] Unhandled Exception: Invalid SkSL:
const float i25 = 0.0003000000142492354;
const int i30 = 0;
const float i33 = 0.6000000238418579;
const int i37 = 1;
const float i42 = 3000.0;
const float i47 = 2000.0;
const float i52 = 0.5;
const float i68 = 1909.0;
const float i79 = 2.0;
const float i98 = 0.0;
const float i108 = 1.0;
const float i228 = 0.22499999403953552;
const float i229 = 3.0;
const float i235 = 0.25;
const float i264 = 3.141590118408203;
uniform float i99;
uniform float2 i86;
float2 i14(float2 i11, float i12, float i13) {
  float i26 = i12 * i25;
  float2 i22 = (i11 + (float2(i26, i26)));
  (i22[i30]) = (((i22[i30]) + ((i33 / i13) * (sin((((i13 * (i11[i37])) + ((i12 / i42) * i13)) + i47))))) + i52);
  (i22[i37]) = (((i22[i37]) + ((i33 / i13) * (sin((((i13 * (i11[i30])) + ((i12 / i42) * i13)) + i68))))) - i52);
  return i22;
}
float4 i20(float2 i18, float2 i19) {
  float i94 = max((i86[i30]), (i86[i37]));
  float2 i78 = ((((float2(i81[0], i81[1])) * i79) - i86) / (float2(i94, i94)));
  float i97 = i98;
  float2 i100 = i78;
  float i102 = i99;
  float i104 = i97;
  i78 = (i14(i100, i102, i104));
  i97 += i108;
  float2 i110 = i78;
  float i112 = i99;
  float i114 = i97;
  i78 = (i14(i110, i112, i114));
  i97 += i108;
  float2 i119 = i78;
  float i121 = i99;
  float i123 = i97;
  i78 = (i14(i119, i121, i123));
  i97 += i108;
  float2 i128 = i78;
  float i130 = i99;
  float i132 = i97;
  i78 = (i14(i128, i130, i132));
  i97 += i108;
  float2 i137 = i78;
  float i139 = i99;
  float i141 = i97;
  i78 = (i14(i137, i139, i141));
  i97 += i108;
  float2 i146 = i78;
  float i148 = i99;
  float i150 = i97;
  i78 = (i14(i146, i148, i150));
  i97 += i108;
  float2 i155 = i78;
  float i157 = i99;
  float i159 = i97;
  i78 = (i14(i155, i157, i159));
  i97 += i108;
  float2 i164 = i78;
  float i166 = i99;
  float i168 = i97;
  i78 = (i14(i164, i166, i168));
  i97 += i108;
  float2 i173 = i78;
  float i175 = i99;
  float i177 = i97;
  i78 = (i14(i173, i175, i177));
  i97 += i108;
  float2 i182 = i78;
  float i184 = i99;
  float i186 = i97;
  i78 = (i14(i182, i184, i186));
  i97 += i108;
  float2 i191 = i78;
  float i193 = i99;
  float i195 = i97;
  i78 = (i14(i191, i193, i195));
  i97 += i108;
  float2 i200 = i78;
  float i202 = i99;
  float i204 = i97;
  i78 = (i14(i200, i202, i204));
  i97 += i108;
  float2 i209 = i78;
  float i211 = i99;
  float i213 = i97;
  i78 = (i14(i209, i211, i213));
  i97 += i108;
  float2 i218 = i78;
  float i220 = i99;
  float i222 = i97;
  i78 = (i14(i218, i220, i222));
  float3 i227 = (float3(i98, ((i228 * (sin((i229 * (i78[i30]))))) + i235), ((i228 * (sin((i229 * (i78[i30]))))) + i228)));
  return (float4((i227[0]), (i227[1]), (i227[2]), i108));
}
half4 main(float2 iFragCoord) {
  float4 i81 = float4(iFragCoord, 0, 0);
  float4 oColor;
  float2 i251 = ((float2(i81[0], i81[1])) / i86);
  float2 i258 = i251;
  float2 i260 = (float2(i81[0], i81[1]));
  oColor = (i20(i258, i260));
  return oColor;
}
SkSL Error:
error: 30: unknown identifier 'i81'
error: 30: unknown identifier 'i81'
error: 32: unknown identifier 'i78'
error: 35: unknown identifier 'i78'
error: 35: unknown identifier 'i100'
error: 37: unknown identifier 'i78'
error: 40: unknown identifier 'i78'
error: 40: unknown identifier 'i110'
error: 42: unknown identifier 'i78'
error: 45: unknown identifier 'i78'
error: 45: unknown identifier 'i119'
error: 47: unknown identifier 'i78'
error: 50: unknown identifier 'i78'
error: 50: unknown identifier 'i128'
error: 52: unknown identifier 'i78'
error: 55: unknown identifier 'i78'
error: 55: unknown identifier 'i137'
error: 57: unknown identifier 'i78'
error: 60: unknown identifier 'i78'
error: 60: unknown identifier 'i146'
error: 62: unknown identifier 'i78'
error: 65: unknown identifier 'i78'
error: 65: unknown identifier 'i155'
error: 67: unknown identifier 'i78'
error: 70: unknown identifier 'i78'
error: 70: unknown identifier 'i164'
error: 72: unknown identifier 'i78'
error: 75: unknown identifier 'i78'
error: 75: unknown identifier 'i173'
error: 77: unknown identifier 'i78'
error: 80: unknown identifier 'i78'
error: 80: unknown identifier 'i182'
error: 82: unknown identifier 'i78'
error: 85: unknown identifier 'i78'
error: 85: unknown identifier 'i191'
error: 87: unknown identifier 'i78'
error: 90: unknown identifier 'i78'
error: 90: unknown identifier 'i200'
error: 92: unknown identifier 'i78'
error: 95: unknown identifier 'i78'
error: 95: unknown identifier 'i209'
error: 97: unknown identifier 'i78'
error: 100: unknown identifier 'i78'
error: 100: unknown identifier 'i218'
error: 101: unknown identifier 'i78'
error: 101: unknown identifier 'i78'
error: 102: unknown identifier 'i227'
error: 102: unknown identifier 'i227'
error: 102: unknown identifier 'i227'
49 errors
#0      FragmentProgram._init (dart:ui/painting.dart:3894:73)
#1      new FragmentProgram._ (dart:ui/painting.dart:3885:5)
#2      FragmentProgram.compile.<anonymous closure> (dart:ui/painting.dart:3872:58)
#3      new Future.<anonymous closure> (dart:async/future.dart:252:37)
#4      _rootRun (dart:async/zone.dart:1418:47)
#5      _CustomZone.run (dart:async/zone.dart:1328:19)
#6      _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
#7      _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1276:23)
#8      _rootRun (dart:async/zone.dart:1426:13)
#9      _CustomZone.run (dart:async/zone.dart:1328:19)
#10     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:1260:23)
#11     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
#12     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:398:19)
#13     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:429:5)
#14     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
```

**Comment by nosmirck** [comment]
Disregard my previous comment!! after some more tinkering with the generated code I found out I'm stupid! I was trying to use `gl_FragCoord` instead of the nicely already provided `vec2 fragCoord` by the fragment method!

All the operations I was using were ok, it's just that using ints is not fully supported yet (the operations like `<` are actually supported for floats)

So, I went back to the original code, replaced `i` and `complexity` with floats and voila! it works now!

This is how the shader looks now:
```glsl
uniform float time;

const float Pi=3.141592653589793;
const float complexity=40.;// More points of color.
const float fluid_speed=3000.;// Drives speed, higher number will make it slower.
const float color_intensity=.225;

vec4 fragment(vec2 uv,vec2 fragCoord){
    vec2 p=(2.*fragCoord.xy-resolution)/max(resolution.x,resolution.y);
    for(float i=1.;i<complexity;i=i+1.){
        
        vec2 newp=p+time*.0003;
        newp.x+=.6/float(i)*sin(i*p.y+time/fluid_speed*(i+2000.))+.5;
        newp.y+=.6/float(i)*sin(i*p.x+time/fluid_speed*(i+1909.))-.5;
        p=newp;
    }
    vec3 col=vec3(0.,color_intensity*sin(3.*p.x)+.25,color_intensity*sin(3.*p.x)+color_intensity);
    return vec4(col,1.);
}
```

**Comment by wolfenrain** [comment]
Yea the integer stuff is a bit of a problem. But I think mapping OP code exceptions to readable/understandable errors would at least make life a bit easier! 

I'll keep this issue open for now until I either add that or write up a good proposal for it. If you have any ideas, be my guest as well for writing a proposal.

**Comment by nosmirck** [comment]
Absolutely! the list you gave me with the supported/unsupported operations helped a lot! otherwise, I would've been in the blank!

I have successfully managed to transcribe a few shaders from the sandbox site and my own old shaders from when I was in university! I'm now going to try my raymarching volume rendering shader that took me ages to write back in 2009 hahahaha

**Comment by wolfenrain** [comment]
Once #34 is merged I'll provide a PR that will allow Umbra to provide more meaningful messages. Got the bare functionality already done
